pkg/codegen/schema/schema.go

```diff
@@ -1510,15 +1546,66 @@ type ResourceSpec struct {
 	Methods map[string]string `json:"methods,omitempty" yaml:"methods,omitempty"`
 }
 
+// ReturnTypeSpec is either ObjectTypeSpec or TypeSpec
+type ReturnTypeSpec struct {
+	ObjectTypeSpec *ObjectTypeSpec
+	TypeSpec       *TypeSpec
+}
+
+// Decoder is an alias for a function that takes (in []byte, out interface{}) and potentially returns an error
+// it is used to abstract json.Unmarshal and yaml.Unmarshal which satisfy this function signature
+type Decoder func([]byte, interface{}) error
+
+func (returnTypeSpec *ReturnTypeSpec) UnmarshalReturnTypeSpec(data []byte, decode Decoder) error {
+	var objectMap map[string]interface{}
+	if err := decode(data, &objectMap); err != nil {
+		return err
+	}
+
+	if len(objectMap) == 0 {
+		return nil
+	}
+
+	var objectSpec *ObjectTypeSpec
+	var typeSpec *TypeSpec
+	if _, hasProperties := objectMap["properties"]; hasProperties {
+		if err := decode(data, &objectSpec); err != nil {
+			return err
+		}
+	} else {
+		if err := decode(data, &typeSpec); err != nil {
+			return err
+		}
+	}
+
+	returnTypeSpec.TypeSpec = typeSpec
+	returnTypeSpec.ObjectTypeSpec = objectSpec
+	return nil
+}
+
+func (returnTypeSpec *ReturnTypeSpec) UnmarshalJSON(inputJSON []byte) error {
+	return returnTypeSpec.UnmarshalReturnTypeSpec(inputJSON, json.Unmarshal)
+}
+
+func (returnTypeSpec *ReturnTypeSpec) UnmarshalYAML(inputYAML []byte) error {
+	return returnTypeSpec.UnmarshalReturnTypeSpec(inputYAML, yaml.Unmarshal)
+}
+
 // FunctionSpec is the serializable form of a function description.
 type FunctionSpec struct {
 	// Description is the description of the function, if any.
 	Description string `json:"description,omitempty" yaml:"description,omitempty"`
 	// Inputs is the bag of input values for the function, if any.
 	Inputs *ObjectTypeSpec `json:"inputs,omitempty" yaml:"inputs,omitempty"`
+	// Determines whether the input bag should be treated as a single argument or as multiple arguments.
+	// When MultiArgumentInputs is non-empty, it must match up 1:1 with the property names in of the Inputs object.
+	// The order in which the properties are listed in MultiArgumentInputs determines the order in which the
+	// arguments are passed to the function.
+	MultiArgumentInputs []string `json:"multiArgumentInputs,omitempty" yaml:"multiArgumentInputs,omitempty"`
 	// Outputs is the bag of output values for the function, if any.
-	Outputs *ObjectTypeSpec `json:"outputs,omitempty" yaml:"outputs,omitempty"`
-	// DeprecationMessage indicates whether or not the function is deprecated.
+	Outputs    *ObjectTypeSpec `json:"outputs,omitempty" yaml:"outputs,omitempty"`
```

Comment:
Deprecation notice - explain why not to specify both `Outputs` and `ReturnType` - i.e. they get written to the same field.